feat(react-router): generic router history

matthiasschwarz opened the pull request:
Changes type of history in router to a generic type parameter to support advanced use cases like providing an own implementation of a `RouterHistory` or to enchance a specific `RouterHistory` with additional features.

schiller-manuel commented:
thanks for the contibution!

please have a look at the build failures and fix those.
please also add a type test (e.g. in packages/react-router/tests/router.test-d.tsx)

matthiasschwarz commented:
> thanks for the contibution!
> 
> please have a look at the build failures and fix those. please also add a type test (e.g. in packages/react-router/tests/router.test-d.tsx)

- Fixed the build issues of the monorepo examples. This is a breaking change though since it was required to add `@tanstack/history` as dependency for the build typescript declarations of the router packages. I'm not sure if this is the correct way to handle that issue.
- The e2e tasks that are still failing are running successfully on my machine.
- Added tests for the history type of the router.

schiller-manuel commented:
@chorobin please have a look about the type changes

schiller-manuel commented:
sorry for the delay
